File: baekjoon/Silver/2178. 미로 탐색/미로 탐색.java
import java.io.*;
import java.util.*;

public class Main {
	static int[] dy = {1, -1, 0 ,0};
	static int[] dx = {0, 0, 1 ,-1};
	
	public static void bfs(int[][] map) {
		Deque<int[]> q = new ArrayDeque<>();
		boolean[][] visited = new boolean[map.length][map[0].length];
		
		q.offer(new int[]{0, 0});
		visited[0][0] = true;
		
		while(!q.isEmpty()) {
			int[] point = q.poll();
			int y = point[0];
			int x = point[1];
			
			for(int i=0; i<4; i++) {
				int ny = y+dy[i];
				int nx = x+dx[i];
				
				if(ny<0 || ny>=map.length || nx<0 || nx>=map[0].length || map[ny][nx]==0 || visited[ny][nx]) continue;
				q.offer(new int[] {ny, nx});
				map[ny][nx] += map[y][x];
				visited[ny][nx] = true;
			}
		}
	}
	
	
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		StringTokenizer st = new StringTokenizer(br.readLine());

		int N = Integer.parseInt(st.nextToken());
		int M = Integer.parseInt(st.nextToken());
		int[][] map = new int[N][M];
		
		for(int i=0; i<N; i++) {
			String str = br.readLine();
			for(int j=0; j<M; j++) {
				map[i][j] = str.charAt(j) - '0';
			}
		}
		
		bfs(map);
		bw.write(Integer.toString(map[N-1][M-1]));
		bw.close();
	}
}

File: baekjoon/Silver/2178. 미로 탐색/미로 탐색.py
"""
[BFS 활용]
1. 전체 maps, visited 선언(N x M)
2. x,y 증감
3. 칸마다 이동하면서 이동 칸수 더해서 갱신
   조건)
       1. maps 크기 벗어나지 않기
       2. maps == 1
       3. visited = False
- 문자열로 maps 받기
"""

import sys
from collections import deque
input = sys.stdin.readline

N, M = map(int, input().split())

maps = [[0]*(M+1) for _ in range(N+1)]
for j in range(N+1):
    if j == 0:  str = '0'*(M+1)
    else:       str = '0' + input()
    for i in range(M+1):
        maps[j][i] = int(str[i])

q = deque()
visited = [[False]*(M+1) for _ in range(N+1)]
q.append((1, 1))
visited[1][1] = True

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]
while q:
    y, x = q.popleft()
    for i in range(4):
        nx = x + dx[i]
        ny = y + dy[i]

        if 1<=nx<M+1 and 1<=ny<N+1 and maps[ny][nx] == 1:
            if not visited[ny][nx]:
                visited[ny][nx] = True
                q.append((ny, nx))
                maps[ny][nx] = maps[y][x] + 1

print(maps[N][M])